pdns/dnsdist-rings.hh

```diff
@@ -0,0 +1,209 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#pragma once
+
+#include <mutex>
+#include <time.h>
+#include <unordered_map>
+
+#include <boost/circular_buffer.hpp>
+#include <boost/variant.hpp>
+
+#include "dnsname.hh"
+#include "iputils.hh"
+
+
+struct Rings {
+  struct Query
+  {
+    struct timespec when;
+    ComboAddress requestor;
+    DNSName name;
+    uint16_t size;
+    uint16_t qtype;
+    struct dnsheader dh;
+  };
+  struct Response
+  {
+    struct timespec when;
+    ComboAddress requestor;
+    DNSName name;
+    uint16_t qtype;
+    unsigned int usec;
+    unsigned int size;
+    struct dnsheader dh;
+    ComboAddress ds; // who handled it
+  };
+
+  struct Shard
+  {
+    boost::circular_buffer<Query> queryRing;
+    boost::circular_buffer<Response> respRing;
+    std::mutex queryLock;
+    std::mutex respLock;
+  };
+
+  Rings(size_t capacity=10000, size_t numberOfShards=1, size_t nbLockTries=5, bool keepLockingStats=false): d_blockingQueryInserts(0), d_blockingResponseInserts(0), d_deferredQueryInserts(0), d_deferredResponseInserts(0), d_nbQueryEntries(0), d_nbResponseEntries(0), d_currentShardId(0), d_numberOfShards(numberOfShards), d_nbLockTries(nbLockTries), d_keepLockingStats(keepLockingStats)
+  {
+    setCapacity(capacity, numberOfShards);
+    if (numberOfShards <= 1) {
+      d_nbLockTries = 0;
+    }
+  }
+  std::unordered_map<int, vector<boost::variant<string,double> > > getTopBandwidth(unsigned int numentries);
+  size_t numDistinctRequestors();
+  void setCapacity(size_t newCapacity, size_t numberOfShards)
+  {
+    if (numberOfShards < d_numberOfShards) {
+      throw std::runtime_error("Decreasing the number of shards in the query and response rings is not supported");
+    }
+
```

Comment:
Maybe make sure the `newCapacity / numberOfShards` ratio is coherent ?

Reply:
This may reduce the size of the indivual rings and  have some strange behaviour regarding `d_nbQueryEntries`

Reply:
Indeed, that's why it's only allowed at configuration time and not at run-time. I could write a few words about that here as well though.